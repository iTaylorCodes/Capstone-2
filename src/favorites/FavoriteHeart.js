import { useContext, useState } from "react";
import UserContext from "../auth/UserContext";
import { NeighborhoodApi } from "../api/api";

/** Favorites List.
 *
 * Heart component for add/removing hotelIds from user favorites.
 *
 * Rendered by HotelCard and FavoriteHotelCard components.
 */
const FavoriteHeart = ({ hotelId }) => {
  let { currentUser } = useContext(UserContext);
  if (!currentUser)
    currentUser = { favoritedProperties: [], notLoggedIn: true };
  const defaultState = currentUser.favoritedProperties.includes(hotelId);
  const [isFavorite, setIsFavorite] = useState(defaultState);
  const [errorMsg, setErrorMsg] = useState("");

  const setFav = async (e) => {
    if (currentUser.notLoggedIn) {
      setErrorMsg("Signup or Login to save your favorite hotels.");
      return;
    }
    await NeighborhoodApi.favoriteProperty(
      currentUser.username,
      parseInt(hotelId, 10)
    );
    setIsFavorite(true);
  };

  const removeFav = async (e) => {
    if (currentUser.notLoggedIn) {
      setErrorMsg("Signup or Login to save your favorite hotels.");
      return;
    }
    await NeighborhoodApi.unFavoriteProperty(
      currentUser.username,
      parseInt(hotelId, 10)
    );
    setIsFavorite(false);
  };

  return (
    <div
      className="FavoriteHeart d-flex justify-content-end"
      style={{ margin: "-5px" }}
    >
      {errorMsg ? (
        <span className="text-danger me-1 fw-light fs-6">{errorMsg}</span>
      ) : (
        ""
      )}
      <i
        className="fa fa-heart"
        style={{ color: isFavorite ? "#0275d8" : "", cursor: "pointer" }}
        onClick={isFavorite ? removeFav : setFav}
        data-testid="favHeart"
      ></i>
    </div>
  );
};

export default FavoriteHeart;
