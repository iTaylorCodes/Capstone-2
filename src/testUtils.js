import UserContext from "./auth/UserContext";

const testUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
  password: "password",
  favoritedProperties: ["12345"],
};

const UserProvider = ({ children, currentUser = testUser }) => (
  <UserContext.Provider value={{ currentUser }}>
    {children}
  </UserContext.Provider>
);

export { UserProvider };
