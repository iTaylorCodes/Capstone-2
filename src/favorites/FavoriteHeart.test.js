import FavoriteHeart from "./FavoriteHeart";
import { render, fireEvent, act } from "@testing-library/react";
import { UserProvider } from "../testUtils";

const hotelId = "12345";

it("renders without crashing", () => {
  render(
    <UserProvider>
      <FavoriteHeart hotelId={hotelId} />
    </UserProvider>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <UserProvider>
      <FavoriteHeart hotelId={hotelId} />
    </UserProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("updates database on click", async () => {
  await act(async () => {
    const { getByTestId } = render(
      <UserProvider>
        <FavoriteHeart hotelId={hotelId} />
      </UserProvider>
    );

    const heart = getByTestId("favHeart");

    expect(heart).toHaveStyle({
      color: "#0275d8",
      cursor: "pointer",
    });

    await fireEvent.click(heart);

    expect(getByTestId("favHeart")).toHaveStyle({
      cursor: "pointer",
    });

    await fireEvent.click(heart);

    expect(getByTestId("favHeart")).toHaveStyle({
      color: "#0275d8",
      cursor: "pointer",
    });
  });
});
