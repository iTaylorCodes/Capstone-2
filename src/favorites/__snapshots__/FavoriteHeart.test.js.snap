// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="FavoriteHeart d-flex justify-content-end"
    style="margin: -5px;"
  >
    <i
      class="fa fa-heart"
      data-testid="favHeart"
      style="color: rgb(2, 117, 216); cursor: pointer;"
    />
  </div>
</DocumentFragment>
`;
